keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,971 @@
+`status.condition` struct. This is to keep parity with the 
+[metav1.Condition struct](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1620-L1625). 
+
+The net result will be a new `status.observedGeneration` field for the kubelet
+to express which generation the top-level status relates to, and a new `status.conditions[i].observedGeneration`
+field for the writer of that condition to express which generation that condition
+relates to.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Because there is only one singular `status.observedGeneration` in the pod status, only
+one writer can set it. This is consistent with other object types that have
+`status.observedGeneration`, and the expectation is that the primary controller for
+the object sets the field. For a pod, that primary controller is the kubelet, so
+once a pod is bound to a node, we expect that the kubelet on that node is the
+sole writer of `status.observedGeneration`. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Today, it is possible for a client to custom-set a pod's `metadata.generation`, but once
+set, the `metadata.generation` cannot be updated. That means that there is a theoretical 
+possibility for an external client to be setting `metadata.generation` on pod 
+create and depend on that fixed value somehow in its own reconciliation logic.
+
+That said, the `metadata.generation` field is described as ["set by the system and monotonically increasing"](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata),
+and thus it should be clear enough that `metadata.generation` was not intended to be
+used in this way.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `metadata.generation` to 1. For any updates
+to the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core),
+and additionally updates to annotations that end up getting copied to the PodSpec
```

Comment:
when saying newly created, this is after all the webhooks execution, correct? Or mutating webhooks will increment the generation if they update some fields?

Reply:
Yeah, it should be after mutating webhooks run.

Reply:
what would be a recommendation for the mutating webhook. If the webhook updates some PodStatus fields on pod status update API call, does it need to set the `observedGeneration`? 

Let's say it wants to. If webhook will decide to get the "fresh" pod and set the `observedGeneration` from it, than the next kublet-initiated pod update runs, and it still has "older" observedGeneration, will API server accept this update? Are there any other potential update-in-the-middle type of situation that may lead to kubelet sending status updates for an older generation of a Pod?

Reply:
> If the webhook updates some PodStatus fields on pod status update API call, does it need to set the observedGeneration?

That sounds really fragile. The Kubelet owns the pod status. If a webhook is mutating the status, there's probably going to be other problems. Unless the webhook is changing pod conditions that it owns, in which case there is a separate ObservedGeneration in the condition.

It's an interesting question of whether API validation should reject observed generations that regress. On the one hand, something went pretty wrong if that happens, so it might be a good safety mechanism. On the other hand, the Kubelet is the source of truth for the pod status, so it should never be prevented from setting the status. I'm leaning towards allowing it (but maybe logging an error).

Reply:
if we will allow decreasing the value of the `observedGeneration`, than some of my questions are going away. However it makes validation quite weak.

If we are not validating it and only throwing error, it needs to be written explicitly in the KEP

Reply:
When updating status, if the incoming update clears observedGeneration, we should probably just preserve the existing value. We've done that for other pod fields IIRC.

Reply:
added this to the kep to clarify

Reply:
> When updating status, if the incoming update clears observedGeneration (sets back to 0), we should probably just preserve the existing value in statusStrategy#PrepareForUpdate. We've done that for other pod fields IIRC.
> 
> I probably wouldn't do additional validation about observedGeneration moving backwards (no other types do that I can see)

This makes sense to me. I'm happy to follow the convention of whatever the other types are doing today.

Reply:
Added to the kep that the API server setting `metadata.generation` will run _after_ the mutating webhooks, that updating status.observedGeneration to decreasing values is permitted in the API validation, and that if the incoming update request clears observedGeneration we will preserve the previously existing value.